src/realm/bplustree.hpp

```diff
@@ -568,6 +566,13 @@ class BPlusTree : public BPlusTreeBase {
         m_root->bptree_traverse(func);
     }
 
+    void split_if_needed()
+    {
+        while (m_root->get_node_size() > REALM_MAX_BPNODE_SIZE) {
```

Comment:
please test more than one iteration of this loop eg. more than 1000 * 1000 elements. I think the second iteration will fail because `split_root()` assumes that the root is a LeafNode but on the second pass they will be all BPlusTreeInner types

Reply:
This is tested more thoroughly when the node size is 4. `get_node_size()` will return the size of the actual node - no matter if this is an inner node or a leaf.

Reply:
Oh, you had the implementation correct all along and I misread it. Thanks for adding the extra tests though!